connect-inject/endpoints_controller.go

```diff
@@ -0,0 +1,429 @@
+}
+
+// deregisterServiceOnAllAgents queries all agents for service instances that have the metadata
+// "k8s-service-name"=k8sSvcName and "k8s-namespace"=k8sSvcNamespace. The k8s service name may or may not match the
+// consul service name, but the k8s service name will always match the metadata on the Consul service
+// "k8s-service-name". So, we query Consul services by "k8s-service-name" metadata, which is only exposed on the agent
+// API. Therefore, we need to query all agents who have services matching that metadata, and deregister each service
+// instance. When querying by the k8s service name and namespace, the request will return service instances and
+// associated proxy service instances.
+// The argument selectivelyDeregister decides whether to deregister *all* service instances or selectively deregister
+// them only if they are not in endpointsAddressesMap. If selectivelyDeregister is false, this function won't check the
+// endpointsAddressMap so that can be nil.
+func (r *EndpointsController) deregisterServiceOnAllAgents(k8sSvcName, k8sSvcNamespace string, selectivelyDeregister bool, endpointsAddressesMap map[string]bool) error {
+
+	// Get all agents by getting pods with label component=client, app=consul and release=<ReleaseName>
+	list := corev1.PodList{}
+	listOptions := client.ListOptions{
+		Namespace: r.ReleaseNamespace,
+		LabelSelector: labels.SelectorFromSet(map[string]string{
+			"component": "client",
+			"app":       "consul",
+			"release":   r.ReleaseName,
+		}),
+	}
+	err := r.Client.List(r.Ctx, &list, &listOptions)
+	if err != nil {
+		r.Log.Error(err, "failed to get agent pods from Kubernetes")
+		return err
+	}
+
+	// On each agent, we need to get services matching "k8s-service-name" and "k8s-namespace" metadata.
+	for _, pod := range list.Items {
+		// Create client for this agent.
+		client, err := r.getConsulClient(pod.Status.PodIP)
+		if err != nil {
+			r.Log.Error(err, "failed to create a new Consul client", "address", pod.Status.PodIP)
+			return err
+		}
+
+		// Get services matching metadata.
+		svcs, err := client.Agent().ServicesWithFilter(fmt.Sprintf(`Meta[%q] == %q and Meta[%q] == %q`, MetaKeyKubeServiceName, k8sSvcName, MetaKeyKubeNS, k8sSvcNamespace))
+		if err != nil {
+			r.Log.Error(err, "failed to get service instances", MetaKeyKubeServiceName, k8sSvcName)
+			return err
+		}
+
+		// Deregister each service instance that matches the metadata.
+		for svcID, serviceRegistration := range svcs {
+			// If we selectively deregister, only deregister if the address is not in the map. Otherwise, deregister
+			// every service instance.
+			if selectivelyDeregister {
+				if _, ok := endpointsAddressesMap[serviceRegistration.Address]; !ok {
+					// If the service address is not in the Endpoints addresses, deregister it.
+					if err = client.Agent().ServiceDeregister(svcID); err != nil {
+						r.Log.Error(err, "failed to deregister service instance", "consul-service-id", svcID)
```

Comment:
It seems a little strange to me to make it a const since it's just a key in the log. Basically I am not sure what I would call the const to have it be more readable than the string "consul-service-id".